### ISSUE TITLE
Adding holidays in preprocessing data

### ISSUE THREAD

**Issue by teojcryan** [epiforecasts/epinowcast#148]
I'm having issues trying to treat holidays as sundays with `enw_preprocess_data()`. According to the documentation, the argument `holidays` should accept a vector of dates indicating when holidays occur, but when I try this, I receive the following error.

```r
nat_germany_hosp <- epinowcast::germany_covid19_hosp[location == "DE"]

retro_nat_germany <- nat_germany_hosp |>
  epinowcast::enw_filter_report_dates(latest_date = "2021-09-01") |>
  epinowcast::enw_filter_reference_dates(include_days = 40)

epinowcast::enw_preprocess_data(retro_nat_germany,
                                max_delay = 20,
                                by = "age_group",
                                holidays = c(as.Date("2021-07-23")))
#> Error: invalid first argument
```

Just out of curiosity, I tried this without the `as.Date`, leaving the date as a string.

```r
nat_germany_hosp <- epinowcast::germany_covid19_hosp[location == "DE"]

retro_nat_germany <- nat_germany_hosp |>
  epinowcast::enw_filter_report_dates(latest_date = "2021-09-01") |>
  epinowcast::enw_filter_reference_dates(include_days = 40)

epinowcast::enw_preprocess_data(retro_nat_germany,
                                max_delay = 20,
                                by = "age_group",
                                holidays = c("2021-07-23"))
#> Error in .checkTypos(e, names_x): Object '2021-07-23' not found amongst date, .group, age_group, location, delay and 4 more
```

**Comment by seabbs** [comment]
Hi @teojcryan thanks for flagging this and providing a helpful `reprex`

So looking at the code it looks like we have a code/documentation mismatch. Previous versions of the code took a vector of dates and applied these to become holiday effects. The current implementation takes a specified `holidays` variable (specified as a character vector) and assigns them to be Sundays. 

It isn't totally clear to me which of those options is the better one. I would guess it is use case dependent. We definitely need too make sure the code and docs agree though. 

See here for the code implementation:

https://github.com/epiforecasts/epinowcast/blob/603782bece7cea70d68287d4481fc0dc08a006b8/R/preprocess.R#L34

**Comment by teojcryan** [comment]
Thanks @seabbs! I'm still unclear about the required input for the current implementation for adding holidays - for this `holidays` variable, is this a character vector indicating TRUE/FALSE for all dates in `obs` or just the nowcast date? 

Personally, I think it's more intuitive for `holidays` to be a vector of dates.

**Comment by pearsonca** [comment]
as promised in exchange for :beer:, working on this nowish.

Planned implementation:
 - support character scalar OR vector coerceable to dates.

**Comment by pearsonca** [comment]
Hmm, taking a step back: what's wanted here?

 1. Something standardized fed into the backend machinery
 2. Something flexible enough for users to present their various flavors of knowledge about holidays.

The standardized thing is, ultimately, the `day_of_week == "Sunday"` for days that are holidays, but that seems internal.

An intermediate standardization seems to be something like a `holiday` logical column.

Regarding using an existing column in the dataset (i.e. support a string argument for holiday indicating column): seems to me that if the user can be convinced to supply a logical column indicating holiday status, they can be told to name that column `holiday` rather than supplying whatever name it has - that expectation is set elsewhere with e.g. requiring a column named `confirm`.

Seem reasonable, however, to enable supplying something more like raw dates indicating what is and isn't a holiday. Figure most cases will be covered by either processing a series of dates OR a series of date ranges.

**Comment by pearsonca** [comment]
Per discussion, go to just vector of stuff coerceable via `as.IDate`.

**Comment by seabbs** [comment]
@teojcryan if you have a chance to look at @pearsonca PR and make sure it makes sense to you/solves your problem that would be really great. If you have any comments on it please feel free to go to Town. Carl is only scary if you have strong opinions on whether Science is engineering or not 😆 .

**Comment by pearsonca** [comment]
people regularly describe me as "friendly".

@teojcryan when sam indicates "look at" here, i would read that as "try out". both because doing is better, but also because i've written mildly dense code (appropriate to the problem imo, but not for everyone) and just browsing it might not do the trick.

how are you playtesting this currently? via `remotes::install_github(...)` or some such?

if you're working with the repo proper, are you familiar with how to fetch branches-not-on-the-repo? e.g. as in [this example](https://stackoverflow.com/questions/5884784/how-to-pull-remote-branch-from-somebody-elses-repo)

**Comment by teojcryan** [comment]
Hi @pearsonca, thanks for your help on this! yes I'm working with the `remotes::install_github()`. I wasn't sure how to fetch branches that are not on the repo initially but that link was good at explaining it!

**Comment by pearsonca** [comment]
> Hi @pearsonca, thanks for your help on this! yes I'm working with the `remotes::install_github()`. I wasn't sure how to fetch branches that are not on the repo initially but that link was good at explaining it!

If you're using `remotes`, then probably easiest to continue that way. Should be able to install over your current snapshot, just specifying the fork (`pearsonca/epinowcast@holiday`) instead of this, plus modification to get `holiday` branch.

(I think that's correct path, but haven't tried it and writing this from phone on train).

**Comment by seabbs** [comment]
Closed by @pearsonca in #151.
